src/client/datascience/raw-kernel/rawJupyterSession.ts

```diff
@@ -200,35 +200,44 @@ export class RawJupyterSession extends BaseJupyterSession {
                 stacktrace
             });
             if (session) {
-                return (session as RawSession).kernelProcess.dispose();
+                return session.kernelProcess.dispose();
             }
         });
     }
 
-    protected setSession(session: RawSession) {
+    protected setSession(session: RawSession | undefined) {
         super.setSession(session);
-
-        // When setting the session clear our current exit handler and hook up to the
-        // new session process
-        if (this.processExitHandler) {
-            this.processExitHandler.dispose();
-            this.processExitHandler = undefined;
+        if (!session) {
+            return;
         }
-        if (session?.kernelProcess) {
-            // Watch to see if our process exits
-            this.processExitHandler = session.kernelProcess.exited(({ exitCode, reason }) => {
-                sendTelemetryEvent(Telemetry.RawKernelSessionKernelProcessExited, undefined, {
-                    exitCode,
-                    exitReason: getTelemetrySafeErrorMessageFromPythonTraceback(reason)
-                });
-                traceError(`Raw kernel process exited code: ${exitCode}`);
-                this.shutdown().catch((reason) => {
```

Comment:
This was wrong, at this point `session !== this.session`, hence we end up shutting down this session and then INotebook.onDidDispose gets fired and everyone thinks the kernel is dead.
But what is happening is we're restarting a kernel and the old IJupyterSession is being disposed & replaced by the new IJupyterSession.

Not easy to replicate the issues as things happen fast. but when its slow, thats when we have failures (as on CI)